### ISSUE TITLE
In a function with a `noundef` return value, replace `ret undef` with `unreachable`

### ISSUE THREAD

**Issue by scottmcm** [llvm/llvm-project#60717]
`noundef` on the return value means that

> If the value representation contains any undefined or poison bits, the behavior is undefined.

so if the function `ret`urns `undef`, that must be [`unreachable`](https://llvm.org/docs/LangRef.html#unreachable-instruction).

Alive2 confirms this is allowed: <https://alive2.llvm.org/ce/z/bW2DoM>
```
define i32 @src() noundef {
%start:
  ret i32 undef
}
=>
define i32 @tgt() noundef {
%start:
  assume i1 0
}
Transformation seems to be correct!
```

But seemingly it doesn't happen today: <https://llvm.godbolt.org/z/79a91o59q>

**Comment by scottmcm** [comment]
@DianQK Will <https://reviews.llvm.org/D144319> perhaps fix this issue as well?

**Comment by DianQK** [comment]
> @DianQK Will https://reviews.llvm.org/D144319 perhaps fix this issue as well?

Yes, it also solves `nonnull` related transformations. But this needs to wait until the C++ standard library fixes this UB.
If you need this optimization to be done soon, I can submit a scene that only considers `noundef`.

@ldionne This UB fix doesn't look like there's been any new progress for about a month now.

**Comment by DianQK** [comment]
https://reviews.llvm.org/D157227
I added a patch to reduce the impact of this UB.

**Comment by DianQK** [comment]
Based on the discussion at [D157227](https://reviews.llvm.org/D157227), we expect to be back in LLVM 18.

cc @nikic @ldionne

**Comment by DianQK** [comment]
I just realized that the simplest example in the issue is not resolved.  lol
